src/Microsoft.Data.SqlClient/tests/ManualTests/AlwaysEncrypted/AKVTests.cs

```diff
@@ -58,11 +58,74 @@ public void TestEncryptDecryptWithAKV()
             DatabaseHelper.ValidateResultSet(sqlDataReader);
         }
 
+        /*
+         This unit test is designed for PR #2618, which addresses an issue where a failed decryption leaves a connection in a bad state  
+         when it is returned to the connection pool. If a subsequent connection is retried it will result in an "Internal connection fatal error",
+         which causes that connection to be doomed, preventing it from being returned to the pool. 
+         Consequently, retrying a new connection will encounter the same decryption error, leading to a repetitive failure cycle.
+
+         The purpose of this unit test is to simulate a decryption error and verify that the connection remains usable when returned to the pool. 
+         It aims to confirm that three consecutive connections will consistently fail with the "Failed to decrypt column" error.
+        */
+        [ConditionalFact(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringSetupForAE), nameof(DataTestUtility.IsAKVSetupAvailable))]
+        public void ForcedColumnDecryptErrorTestShouldFail()
+        {
+            SqlConnectionStringBuilder builder = new(DataTestUtility.TCPConnectionStringHGSVBS)
+            {
+                ColumnEncryptionSetting = SqlConnectionColumnEncryptionSetting.Enabled,
+                AttestationProtocol = SqlConnectionAttestationProtocol.NotSpecified,
+                EnclaveAttestationUrl = ""
+            };
+
+            // Setup record to query
+            using (SqlConnection sqlConnection = new(builder.ConnectionString))
+            {
+                sqlConnection.Open();
+                Customer customer = new(88, "Microsoft2", "Corporation2");
+
+                using (SqlTransaction sqlTransaction = sqlConnection.BeginTransaction())
+                {
+                    DatabaseHelper.InsertCustomerData(sqlConnection, sqlTransaction, _akvTableName, customer);
+                    sqlTransaction.Commit();
+                }
+            }
+
+            // Setup Empty key store provider
+            Dictionary<String, SqlColumnEncryptionKeyStoreProvider> emptyKeyStoreProviders = new()
+            {
+                { "AZURE_KEY_VAULT", new EmptyKeyStoreProvider() }
+            };
+
+            // Three consecutive connections should fail with "Failed to decrypt column" error. This proves that an error in decryption
+            // does not leave the connection in a bad state.
+            for (int i = 0; i < 3; i++)
```

Comment:
Could you explain what happens internally during each retry and why it specifically requires 3 attempts, and include the explanation in the code?

Reply:
Before the fix was implemented, this is what happens internally:

1.  A connection is used from a connection in pool.
2. That connection is used to query a table with a column that is encrypted.
3. When the decryption failed for some reason, a "Failed to decrypt column." error is raised. The connection is returned to the connection pool with the error still in its buffer.
4. When the client connects again, a connection is used from the pool again and grabbing the same connection recently returned back to the pool.
5. This time when the connection is opened it will raise an "Internal connection fatal error." because that connection still has an existing error in it.  Thus, the driver dooms that connection and not returned it back to the connection pool.
6. When the client connects the third time, it will grab an unused connection from the pool. 
7. That new connection is used to query a table with a column that is encrypted.
8. When the decryption failed for some reason, a "Failed to decrypt column." error is raised. The connection is returned to the connection pool with the error still in its buffer.
9. And the cycle continues.

So, this unit test will try to connect 3 times and do the same query to ensure that the expected error messages are identical each time, "Failed to decrypt column". This proves that the fix worked, and the connection is returned to the pool in a good state.